A3-channel-tracing.md

```diff
@@ -0,0 +1,204 @@
+Channel Tracing
+----
+* Author(s): Noah Eisen (ncteisen@google.com)
+* Approver: markdroth, ctiller
+* Status: Draft
+* Implemented in: N/A
+* Last updated: 03/09/17
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/WFDj3KeHYTI
+
+## Table of Contents
+
+  * [Abstract](#abstract)
+  * [Background](#background)
+  * [Proposal](#proposal)
+  * [Rationale](#rationale)
+  * [Implementation](#implementation)
+     * [Basic Design](#basic-design)
+     * [Tracer Functions](#tracer-functions)
+     * [API Changes](#api-changes)
+     * [Format of Exported Data](#format-of-exported-data)
+     * [Garbage Collection](#garbage-collection)
+
+## Abstract
+
+The document proposes adding a dedicated, in-memory tracing object per channel and subchannel. The tracing object will record important steps in the life of a channel, like address resolution, subchannel creation, channel state changes, etc etc. This trace will be made available with a new gRPC API function call. The long term goal is to incorporate the trace into grpcz or a custom channel tracing UI.
+
+## Background
+
+Channel connectivity issues are the root cause of a significant portion of user reported gRPC bugs (for example, [here](https://groups.google.com/a/google.com/forum/?utm_medium=email&utm_source=footer#!msg/grpc-users/gPFYbRAex1A/VM7h5FWAAgAJ) and [here](https://groups.google.com/a/google.com/forum/#!topic/grpc-users/0bsOqrxYvRc)). At the moment, gRPC has an extensive logging/tracing system in place. However, when juggling hundreds or thousands of connections, the logs can grow too huge to be quickly useful when debugging channel connectivity issues. Often times it is difficult to determine what events led to changes in connectivity.
+
+## Proposal
+
+To implement channel tracing, we will introduce a dedicated, in-memory tracing object that is owned by each channel and subchannel. We will add a new function to the grpc channel api that allows library users to retrieve all of the trace for the channel as a JSON or protobuf object. This returned trace will eventually be hooked into a UI, or exported through the ongoing grpcz project, where it could be used for debugging channel issues. The tracer will be enabled or disabled via a new channel argument, GRPC_ARG_CHANNEL_TRACING.
+
+### Format of Exported Data
+
+The trace data will be exported as JSON from the C-core. However, it will be JSON that is of the correct form to be transformed to protobuf and back again. The JSON schema will be as follows (note that all time strings will follow [RFC 3339](https://tools.ietf.org/html/rfc3339)):
+
+```
+{
+  "channelData": {
+    "numNodesLogged": number,
+    "startTime": timestamp string,
+    "nodes": [
+      {
+        "data": data,
+        "error": string,
+        "time": timestamp string,
+        // can only be one of "GRPC_CHANNEL_INIT", etc etc 
+        "state": enum string,
+        // id of referenced subchannel
+        "subchannelId": number
```

Comment:
If you take my suggestion below about giving all tracers an ID, then this field (minus the "subchannel" prefix) will be needed for subchannels below as well.